src/game_objects/essence.cpp

```diff
@@ -0,0 +1,137 @@
+#include "essence.h"
+#include "events.h"
+#include "asset_loader.h"
+#include "player.h"
+#include "util/util_bt.h"
+#include "timer.h"
+#include "util/util.h"
+#include "game/collectibles/gold.h"
+#include "game/collectibles/nothing.h"
+
+#include <iostream>
+
+Essence::Essence(int id) : GameObject(id){
+    tag = "ESSENCE";
+
+    // Set Value
+    value = std::make_unique<collectibles::Gold>(DEFAULT_ESSENSE_VAL);
+
+    if (id == ON_SERVER) {
+        // Setup rigid body
+        events::RigidBodyData rigid;
+        rigid.object = this;
+        rigid.shape = hit_sphere;
+        rigid.mass = 1;
+        events::add_rigidbody_event(rigid);
+        notify_on_collision = true;
+
+        // Lock y-axis
+        rigidbody->setLinearFactor(btVector3(1, 0.0f, 1));
+
+        random_push();
+    }
+    else {
+        // Make the essence change colors continuously
+        int num_steps = 200;
+        int count = (int) Util::random(0, num_steps * 2);
+        cancel_rainbow = Timer::get()->do_every(
+            std::chrono::milliseconds(10),
+            [&, num_steps, count]() mutable{
+            const float frequency = (2 * glm::pi<float>()) / num_steps;
+            float r, g, b;
+
+            // For smooth looping
+            int i = count;
+            if (count > num_steps)
+                i = num_steps * 2 - count;
+
+            // Adjust rgb values based on frequencies, the last number dictates the 'balance' (0 - 255)
+            r = sinf(frequency * i + 0) * 127 + 200;
+            g = sinf(frequency * i + 2) * 127 + 200;
+            b = sinf(frequency * i + 4) * 127 + 200;
+
+            set_filter_color({ r, g, b });
+
+            count = (count < num_steps * 2) ? count + 1 : 0;
+        });
+    }
+}
+
+void Essence::s_update_this() {
+
+    // Get the transform from Bullet and update
+    btTransform t;
+    rigidbody->getMotionState()->getWorldTransform(t);
+    btVector3 to_set = t.getOrigin();
+    transform.set_position(util_bt::convert_vec3(to_set));
+}
+
+void Essence::s_on_collision(GameObject *other) {
+    Player *player = dynamic_cast<Player*>(other);
+    if (player && !player->is_invulnerable()) {
+        value->collected_by(player); // Award player
+        value = std::make_unique<collectibles::Nothing>();
+
+        events::dungeon::network_collision_event(player->get_id(), id);
+
+        // Call disable() once client side animation ends
+        events::disable_rigidbody_event(this); // Disables rigid body first to prevent moving
+        Timer::get()->do_after(
+            std::chrono::milliseconds(500),
+            [&]() {
+            disable();
```

Comment:
Yeah, ideally we would delete it. But I don't think our codebase allows us to actually delete objects safely, cause it'd cause a bunch of null references in various places since none of our game objects actually have destructors. I'd rather have that done and tested in a separate PR.